File: python/torrent/lex.py
import element

ET_UNDEF = 0
ET_INTEGER = 1
ET_STRING = 2
ET_LIST = 3
ET_DICT = 4


class Lex:
    def __init__(self, fp):
        self.fp = fp

    @staticmethod
    def decode(b):
        if b == '':
            return ''
        return b.decode('latin-1')

    def peek(self):
        c = self.fp.read(1)
        self.fp.seek(-1, 1)
        return self.decode(c)

    def getch(self):
        return self.decode(self.fp.read(1))

    def gettok(self):
        c = self.peek()
        if c == 'd':
            self.getch()
            return ET_DICT
        elif c == 'i':
            self.getch()
            return ET_INTEGER
        elif c == 'l':
            self.getch()
            return ET_LIST
        elif c.isdigit():
            return ET_STRING
        else:
            self.getch()
            return c

    def dictionary(self):
        e = element.Dictionary()
        c = self.peek()
        while c != 'e' and c != '':
            k = self.string()
            v = self.getelement()
            e[k.value] = v
            c = self.peek()

        self.gettok()  # 'e'

        return e

    def integer(self):
        e = element.Integer()
        x = ""
        c = self.peek()
        while c != 'e' and c != '':
            x += self.getch()
            c = self.peek()

        self.gettok()  # 'e'
        e.value = int(x)

        return e

    def list(self):
        e = element.List()
        c = self.peek()
        while c != 'e' and c != '':
            e.value.append(self.getelement())
            c = self.peek()

        self.gettok()  # 'e'

        return e

    def string(self):
        e = element.String()
        x = ""
        cnt = 0
        c = self.peek()
        while c != ':' and c != '':
            x += self.getch()
            c = self.peek()

        length = int(x)
        self.gettok()  # ':'

        while cnt < length:
            c = self.getch()
            if c == '':
                break
            e.value += c
            cnt += 1

        return e

    def getelement(self):
        t = self.gettok()
        if t == ET_DICT:
            return self.dictionary()
        elif t == ET_INTEGER:
            return self.integer()
        elif t == ET_LIST:
            return self.list()
        elif t == ET_STRING:
            return self.string()
        else:
            raise Exception("undefined token \"%s\"." % t)

File: c/torrent/lex.c
/*
 * LEX.C
 *
 * Lexical analyzer for bencoded files
 * Copyright (c) 2006, Thomas A. Rieck
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <memory.h>
#include <time.h>

#include "bencode.h"
#include "btdict.h"
#include "lex.h"
#include "ui.h"

static int gettok(FILE *fp);
static int lookahead(FILE *fp);
static bt_element dictionary(FILE *fp);
static bt_element integer(FILE *fp);
static bt_element list(FILE *fp);
static bt_element string(FILE *fp);

/* get next input token from stream */
int gettok(FILE *fp)
{
	int c;
	c = fgetc(fp);

	switch (c) {
	case 'd':
		return ET_DICT;
	case 'i':
		return ET_INTEGER;
	case 'l':
		return ET_LIST;
	default:
		if (isdigit(c)) {
			ungetc(c, fp);
			return ET_STRING;
		}
	}
	return c;
}

/* lookahead one token */
int lookahead(FILE *fp)
{
	int c = fgetc(fp);
	ungetc(c, fp);
	return c;
}

/* load element from stream */
bt_element getelement(FILE *fp)
{
	bt_element e;
	int t;

	memset(&e, 0, sizeof(bt_element));
	t = gettok(fp);

	switch (t) {
	case ET_DICT:
		return dictionary(fp);
	case ET_INTEGER:
		return integer(fp);
	case ET_LIST:
		return list(fp);
	case ET_STRING:
		return string(fp);
	default:
		ui_error("unexpected token 0x%.2x.\n", t);
		break;
	}

	return e;
}

/* load dictionary from stream */
bt_element dictionary(FILE *fp)
{
	bt_element e, k, v;
	int c;

	memset(&e, 0, sizeof(bt_element));
	e.type = ET_DICT;
	e.dval = alloc_bt_dict();

	while ((c = lookahead(fp)) != 'e' && c != EOF) {
		k = string(fp);
		v = getelement(fp);
		bt_dict_insert(e.dval, k.sval->data, &v);
		free_element(k);
	}

	gettok(fp);	/* 'e' */

	return e;
}

/* load integer from stream */
bt_element integer(FILE *fp)
{
	bt_element e;
	memset(&e, 0, sizeof(bt_element));

	e.type = ET_INTEGER;
	fscanf(fp, "%I64de", &e.ival);

	return e;
}

/* load list from stream */
bt_element list(FILE *fp)
{
	bt_element e;
	bt_list *l = NULL;
	int c;

	memset(&e, 0, sizeof(bt_element));
	e.type = ET_LIST;

	while ((c = lookahead(fp)) != 'e' && c != EOF) {
		if (l == NULL) e.lval = l = alloc_bt_list();
		else {
			l->next = alloc_bt_list();
			l = l->next;
		}
		l->e = getelement(fp);
	}

	gettok(fp);	/* 'e' */

	return e;
}

/* load string from stream */
bt_element string(FILE *fp)
{
	bt_element e;
	int c, len = 0;
	char *p;

	e.type = ET_STRING;
	fscanf(fp, "%d:", &len);
	e.sval = alloc_bt_string();
	e.sval->nlen = len;
	e.sval->data = p = (char*)malloc(len + 1);

	while (p - e.sval->data < len) {
		c = fgetc(fp);
		if (c == EOF) break;
		*p++ = c;
	}
	*p = '\0';

	return e;
}
